Remove hyper & tokio; implement an HTTP server; and add system tests.

ltratt opened the pull request:
This commit is a thorough yak shave. First I tried to upgrade hyper, which we previously used for snare's HTTP server. However, it has thoroughly changed its API and no longer provides an HTTP server directly (there is an external crate which is supposed to work, but I couldn't get it to compile). That slowly made me rethink snare's HTTP server, because the use of hyper is why snare also required tokio, with all that comes with it (async, lots of dependencies, and the overloading of my pea brain).

So, second, I removed hyper and tokio and implemented my own minimalistic HTTP server (heavily modelled on pizauth). This hugely reduces our dependencies (from 213 -> 172 for `cargo build`) and reduces the final binary size by 20% in release mode.

However, how do we know that taking on the burden of running our own HTTP server has left snare correct? One of snare's long-term weaknesses is that we haven't have good system-level tests, and I've often been lacking confidence when working on snare that my changes haven't broken something important.

So, third, and finally, this commit adds system level tests is `tests/basic.rs`. The most thorough of these fires a full GitHub-esque request at snare, forcing it to authenticate, and run a test command. While it's not perfect, it's a lot better than nothing.

This commit has also survived being run against GitHub itself. There is more to do (e.g. the tests currently bind to fixed ports; and there aren't enough system tests), but this feels like a sufficiently big unit of work in and of itself.

ltratt committed:
Remove hyper & tokio; implement an HTTP server; and add system tests.

This commit is a thorough yak shave. First I tried to upgrade hyper,
which we previously used for snare's HTTP server. However, it has
thoroughly changed its API and no longer provides an HTTP server
directly (there is an external crate which is supposed to work, but I
couldn't get it to compile). That slowly made me rethink snare's HTTP
server, because the use of hyper is why snare also required tokio,
with all that comes with it (async, lots of dependencies, and the
overloading of my pea brain).

So, second, I removed hyper and tokio and implemented my own
minimalistic HTTP server (heavily modelled on pizauth). This hugely
reduces our dependencies (from 213 -> 172 for `cargo build`) and reduces
the final binary size by 20% in release mode.

However, how do we know that taking on the burden of running our own
HTTP server has left snare correct? One of snare's long-term weaknesses
is that we haven't have good system-level tests, and I've often been
lacking confidence when working on snare that my changes haven't broken
something important.

So, third, and finally, this commit adds system level tests is
`tests/basic.rs`. The most thorough of these fires a full GitHub-esque
request at snare, forcing it to authenticate, and run a test command.
While it's not perfect, it's a lot better than nothing.

This commit has also survived being run against GitHub itself. There is
more to do (e.g. the tests currently bind to fixed ports; and there
aren't enough system tests), but this feels like a sufficiently big unit
of work in and of itself.

vext01 commented on src/httpserver.rs:
```diff
@@ -1,178 +1,228 @@
+    let listener = TcpListener::bind(snare.conf.lock().unwrap().listen)?;
+    let active = Arc::new(AtomicUsize::new(0));
+    for mut stream in listener.incoming().flatten() {
+        while active.fetch_add(1, Ordering::Relaxed) > MAX_SIMULTANEOUS_CONNECTIONS {
```
While adding one is greater than what we are permitting to accept, subtract one... This is clever in a way I'm not sure what it does, and needs a comment I reckon!

ltratt replied:
Hopefully addressed in 36f8f5e.

vext01 replied:
I think I get it.

vext01 commented on src/httpserver.rs:
```diff
@@ -1,178 +1,228 @@
-        .map(|s| s.to_owned())
+/// A very literal, and rather unforgiving, implementation of RFC2616 (HTTP/1.1), returning the URL
+/// of GET requests: returns `Err` for anything else.
+fn parse_get(stream: &mut TcpStream) -> Result<(HashMap<String, String>, Vec<u8>), Box<dyn Error>> {
```
This function is complicated and is bound to have bugs. It could benefit from unit testing.

Perhaps `TcpStream` is `Read` and this could be generalised to accepting `&str`s and `TcpStream`s, allowing easier testing?

ltratt replied:
It can't accept `&str`s because it has to do some interpretation (notably of `Content-Length`) before carrying on reading. [My first version didn't do that, but github's HTTP request library doesn't do "the right thing" if you try and simplify this!]

The good news is that the new system tests do partially test this. Not completely, of course, and I want to add more tests, but it is at least somewhat tested.

vext01 replied:
I suppose in a way, by connecting to localhost and writing a string down a socket, you are somewhat accepting a string anyway.

I won't push, but it'd be good to test this function in isolation, if not now, later.

ltratt replied:
It's worth stating that the only way to test this function is by passing it data over a socket. The way it has to work is first read the headers and no more. Interpret the headers. If there's `Content-Length`, read *exactly* that number of bytes, no more, no less (this is not theoretical: if you don't do this, GitHub's requests won't work!). So it has to read from a stream, not a string.

vext01 replied:
sure, but you could split the tcp reading from the logic. The logic could accept a tuple of `headers: &str, data: &str`, then you can test the inner.

ltratt replied:
It's worse than that: you *have* to parse the headers before you try reading the body from the network. [This was news to me, FWIW, and cost me a couple of hours debugging yesterday!]

ltratt commented on tests/basic.rs:
```diff
@@ -0,0 +1,179 @@
+}
+
+#[test]
+fn full_request() -> Result<(), Box<dyn Error>> {
```
Agreed. My aim is for the next PR to add several "negative" tests like this.

vext01 commented:
Please squash.

ltratt commented:
Squashed.

ltratt committed:
Default to `/bin/sh` if `$SHELL` isn't set.

To my surprise, docker doesn't seem to set this variable by default, but
I don't think I should have been surprised: we're not exactly running
under a shell. Defaulting to `/bin/sh` seems the right thing to do in
such a situation.

ltratt commented:
@vext01 I *think* the problem might have been simple that docker doesn't set `$SHELL` (which, on reflection, is a reasonable thing for it not set). https://github.com/softdevteam/snare/pull/122/commits/18c887d95ee99b52e8d8b1c0d08278c3f4ea4641 fixes that (and I think should remain a separate commit). This works for me locally now, and is probably worth trying to merge as-is if you're OK with it.
